### PR TITLE
(#16) Add Directory Safety Checks

### LINKED ISSUE #16: Add Safety Checking
Safety Checking
---

 * [x] Directories Exist?
 * [ ] Nil values are absent from Vagrant output
 * [x] Check methods are reusable as mix-ins

nil values in config
---

Functional testing in #11 and #13 uncovered the following behavior.  When configuration keys are missing, the program operates without error, but the resulting Vagrantfile is bad.  For example, given the minimal configuration files in the functional test in 69a36bf6ff14f2e081a2b829f762ad4964d8c188 of #13, the following Vagrantfile is produced.  Note the nil values:

```
rzo generate is expected to produce a Vagrantfile with one VM defined
=====================================================================
# This file generated with Rizzo 0.3.0 on 2017-09-01 18:37:20 -0700
# https://github.com/ghoneycutt/rizzo
Vagrant.configure(2) do |config|
  # use 'vagrant plugin install vagrant-proxyconf' to install
  if Vagrant.has_plugin?('vagrant-proxyconf')
    config.proxy.http  = ENV['HTTP_PROXY']  if ENV['HTTP_PROXY']
    config.proxy.https = ENV['HTTPS_PROXY'] if ENV['HTTPS_PROXY']
  end

  config.vm.define "puppetca", autostart: false do |cfg|
    cfg.vm.box = nil
    cfg.vm.box_url = nil
    cfg.vm.box_download_checksum = nil
    cfg.vm.box_download_checksum_type = nil
    cfg.vm.provider :virtualbox do |vb|
      vb.customize ['modifyvm', :id, '--memory', nil]
    end
    cfg.vm.hostname = nil
    cfg.vm.network 'private_network',
      ip: nil,
      netmask: nil
    cfg.vm.synced_folder "/var/folders/d1/94j28v2j1sdc6n_s6knb0kf00000gn/T/tmp.iOTdXdQW/git/puppetdata", "/repos/puppetdata",
      owner: "root", group: "root"
    cfg.vm.synced_folder "/Users/jeff/git/ghoneycutt-modules", "/repos/ghoneycutt",
      owner: "root", group: "root"
    config.vm.synced_folder "/Users/jeff/git/bootstrap",
      nil,
      owner: 'vagrant', group: 'root'
    config.vm.provision 'shell', inline: "echo 'modulepath = ./modules:./puppetdata/modules:./ghoneycutt/modules' > /environment.conf"
    config.vm.provision 'shell', inline: "/bin/bash / "
  end
end
# -*- mode: ruby -*-
# vim:ft=ruby
```

Inputs are a control repo JSON of:

```
{"nodes":[{"name":"puppetca"}]}
```

And a `~/.rizzo.json` of:

```
{
  "defaults": {
    "bootstrap_repo_path": "/Users/jeff/git/bootstrap"
  },
  "control_repos": [
    "/var/folders/d1/94j28v2j1sdc6n_s6knb0kf00000gn/T/tmp.iOTdXdQW/git/puppetdata",
    "/Users/jeff/git/ghoneycutt-modules"
  ],
  "puppetmaster": {
    "name": [
      "puppetca",
      "puppet"
    ],
    "modulepath": [
      "./modules",
      "./puppetdata/modules",
      "./ghoneycutt/modules"
    ],
    "synced_folders": {
      "/repos/puppetdata": {
        "local": "/var/folders/d1/94j28v2j1sdc6n_s6knb0kf00000gn/T/tmp.iOTdXdQW/git/puppetdata",
        "owner": "root",
        "group": "root"
      },
      "/repos/ghoneycutt": {
        "local": "/Users/jeff/git/ghoneycutt-modules",
        "owner": "root",
        "group": "root"
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by jeffmccune**
Without this patch the resulting Vagrantfile may be invalid, or rzo may be
configured to read from directories which do not exist.  This patch addresses
the problem by adding safety checking methods.

Closes #16

### CODE REVIEW

**Comment by jeffmccune** [comment]
Work in progress, up for early review.  Need to update the functional tests and add checking for nil values in the output Vagrantfile.  Checking of the input directories is implemented.

**Comment by ghoneycutt** [comment]
Looks good! Perhaps we should use a json schema?

http://json-schema.org/
https://github.com/ruby-json-schema/json-schema

**Comment by jeffmccune** [comment]
> Looks good! Perhaps we should use a json schema?

That's a great idea, looking into it today.

File: lib/rzo/app/config_validation.rb [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+      # git directory.
+      #
+      # Return nil if all validations pass, Otherwise return the Issue found.
+      def validate_existence(path)
+        pn = Pathname.new(path)
+        git = pn + '.git'
+        return Issue.new("#{pn} is not an absolute path.  It must be fully qualified, not relative") unless pn.absolute?
+        return Issue.new("#{pn} is not a directory.  Has it been cloned?") unless pn.directory?
+        return Issue.new("#{pn} is not readable.  Are permissions correct?") unless pn.readable?
+        return Issue.new "#{git} does not exist.  Has #{git.dirname} been coloned properly?" unless git.directory?
```
Comment (ghoneycutt) [inline lib/rzo/app/config_validation.rb]: spelling: cloned

File: lib/rzo/app/config_validation.rb [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      end
+
+      ##
+      # Validate the bootstrap repo path exists.
+      def validate_bootstrap_repo_path(config)
+        validate_existence(config['defaults']['bootstrap_repo_path'])
```
Comment (ghoneycutt) [inline lib/rzo/app/config_validation.rb]: perhaps we should define an array of keys to check with validate_existence?

File: README.md [hunk 3]
```diff
@@ -4,6 +4,11 @@ Rizzo is a heavily customized Vagrant configuration and work flow with a
 role based focus. It is meant to make working with Vagrant easier and
 purpose built for layered Puppet control repositories.
 
+Rizzo loads a personal configuration file, `~/.rizzo.json` by default, which
+lists one or more control repositories.  Rizzo then looks for a loads a
+`.rizzo.json` configuration file located at the root of each listed control
```
Comment (ghoneycutt) [inline README.md]: it only loads from the top layer control repo

Comment (jeffmccune) [reply README.md]: Right, I keep thinking it merges them for some reason.

File: README.md [hunk 4]
```diff
@@ -124,11 +133,16 @@ Change the paths to your git repos
 }
 ```
 
-Once you have `~/.rizzo.json`, change to your top level control repository and generate your `Vagrantfile`.
+Once you have a personal config file, `~/.rizzo.json`, change directories to
+your top level control repository and generate your `Vagrantfile`:
 
 ```shell
+cd ~/git/puppetdata
```
Comment (ghoneycutt) [inline README.md]: `puppet-control-myteam` instead of puppetdata as we want to encourage team's to have their own and layer on top of each other.

File: README.md [hunk 5]
```diff
@@ -155,8 +169,8 @@ VM, run `vagrant status NAME`.
 ### defaults
 
 The defaults hash is merged with each node entries hash. Put user
-specific entries in `~/.rizzo.json` and project specific entries in
-`${PATH_TO_CONTROL_REPO}/.rizzo.json`.
+specific entries in the personal configuration file at `~/.rizzo.json` and
+project specific entries in `${PATH_TO_CONTROL_REPO}/.rizzo.json`.
```
Comment (ghoneycutt) [inline README.md]: s/project/control repo/

File: lib/rzo/app/config_validation.rb (hunk 2 above)
Comment (jeffmccune) [reply lib/rzo/app/config_validation.rb]: Good idea, get's a little tricky since they're nested but I'll definitely look into it.

File: lib/rzo/app/config_validation.rb (hunk 1 above)
Comment (jeffmccune) [reply lib/rzo/app/config_validation.rb]: Fixed

File: README.md (hunk 4 above)
Comment (jeffmccune) [reply README.md]: Fixed, added a third entry in the example above to match up.

File: README.md (hunk 5 above)
Comment (jeffmccune) [reply README.md]: Fixed

**Comment by jeffmccune** [comment]
@ghoneycutt This is good to go.  Could use more testing, but am interested in getting this out and adding it later.

File: README.md [hunk 6]
```diff
@@ -4,6 +4,12 @@ Rizzo is a heavily customized Vagrant configuration and work flow with a
 role based focus. It is meant to make working with Vagrant easier and
 purpose built for layered Puppet control repositories.
 
+Rizzo loads a personal configuration file, `~/.rizzo.json` by default, which
+lists one or more control repositories.  Rizzo then looks for a loads a
```
Comment (ghoneycutt) [inline README.md:8]: looks for and loads

File: scripts/functional_gem_behavior.sh [hunk 7]
```diff
@@ -232,24 +264,14 @@ else
...
-  echo "STDOUT is expected to be empty, got:"
-  cat $stdout
-  fail "rzo roles produced STDOUT output."
-else
-  pass "STDOUT is zero bytes"
-fi
+
+match "warns about puppetdata not being a directory" 'WARN .*puppetdata is not a directory' $stderr
+match "warns about ghoneycutt-modules not being a directory" 'WARN .*ghoneycutt-modules is not a directory' $stderr
+match "warns about bootstrap not being a directory" 'WARN .*bootstrap is not a directory' $stderr
```
Comment (ghoneycutt) [inline scripts/functional_gem_behavior.sh:274]: after this run mkdir on the three directories and run rzo again. Now no warnings should be present.

File: scripts/functional_gem_behavior.sh [hunk 8]
```diff
@@ -232,24 +264,14 @@ else
...
-  pass "STDERR is zero bytes"
-fi
-if [[ -s $stdout ]]; then
-  echo "STDOUT is expected to be empty, got:"
-  cat $stdout
-  fail "rzo roles produced STDOUT output."
-else
-  pass "STDOUT is zero bytes"
-fi
+
```
Comment (ghoneycutt) [inline scripts/functional_gem_behavior.sh:271]: Since each schema validation has its own warnings, we should have an empty .rizzo.json with just `{}` and see that all the expected warnings happen.

**Comment by ghoneycutt** [comment]
Released in v0.4.0
